File: src/models/health.mjs
import { UPSTREAM_MODULE, TARGET_HOST_VAL, TARGET_PORT_INT, AGENT } from "../config.mjs";
import { log } from "../logger.mjs";

const PROBE_INTERVAL = 60000;     // recovery probe every 60s
const PROBE_TIMEOUT = 8000;       // probe deadline
const BASE_COOLDOWN = 30000;      // 30s first offence
const MAX_COOLDOWN = 600000;      // 10min max lock

const failedUntil = new Map();    // model_id → timestamp
const failCounts = new Map();     // model_id → consecutive failure count

const BACKOFF = [30000, 60000, 120000, 300000, 600000]; // 30s, 1m, 2m, 5m, 10m

export function isModelHealthy(modelId) {
  const ts = failedUntil.get(modelId);
  if (!ts) return true;
  return Date.now() > ts;
}

export function getHealthyModels(models) {
  return models.filter((m) => isModelHealthy(m.id));
}

export function markModelFailed(modelId, statusCode) {
  if (!modelId) return;
  if (statusCode >= 500) {
    const count = (failCounts.get(modelId) || 0) + 1;
    failCounts.set(modelId, count);
    const idx = Math.min(count - 1, BACKOFF.length - 1);
    const cooldown = BACKOFF[idx];
    failedUntil.set(modelId, Date.now() + cooldown);
    const ts = new Date().toISOString();
    log(ts, `MODEL UNHEALTHY: ${modelId} (${statusCode}, #${count}) — locked for ${cooldown / 1000}s`);
  }
}

export function markModelExhausted(modelId) {
  if (!modelId) return;
  failedUntil.set(modelId, Date.now() + 120000);
  const ts = new Date().toISOString();
  log(ts, `MODEL EXHAUSTED (429): ${modelId} — locked for 120s`);
}

export function markModelDegraded(modelId, reason = "degraded") {
  if (!modelId) return;
  failedUntil.set(modelId, Date.now() + 60000);
  const ts = new Date().toISOString();
  log(ts, `MODEL DEGRADED: ${modelId} (${reason}) — locked for 60s`);
}

function clearModelLock(modelId) {
  failedUntil.delete(modelId);
  failCounts.delete(modelId);
  log(new Date().toISOString(), `MODEL RECOVERED: ${modelId}`);
}

async function probeModel(modelId, wafCookie, spoofHeaders) {
  return new Promise((resolve) => {
    const headers = { ...spoofHeaders, "Content-Type": "application/json" };
    if (wafCookie) headers["Cookie"] = wafCookie;

    const body = JSON.stringify({
      model: modelId, max_tokens: 1, stream: false,
      messages: [{ role: "user", content: "." }],
    });

    const req = UPSTREAM_MODULE.request(
      {
        hostname: TARGET_HOST_VAL,
        port: TARGET_PORT_INT,
        path: "/v1/messages",
        method: "POST",
        headers,
        agent: AGENT,
        rejectUnauthorized: true,
        timeout: PROBE_TIMEOUT,
      },
      (res) => {
        res.resume();
        res.on("end", () => resolve(res.statusCode === 200));
        res.on("error", () => resolve(false));
      }
    );
    req.on("error", () => resolve(false));
    req.on("timeout", () => { req.destroy(); resolve(false); });
    req.write(body);
    req.end();
  });
}

let probeTimer = null;

export function startProbeLoop(getModelsFn, getWafFn, spoofHeaders) {
  if (probeTimer) return;
  probeTimer = setInterval(async () => {
    if (failedUntil.size === 0) return;
    for (const [modelId, until] of failedUntil) {
      if (Date.now() > until) {
        const ok = await probeModel(modelId, getWafFn(), spoofHeaders);
        if (ok) {
          clearModelLock(modelId);
        } else {
          // Recovery probe still failing — extend lock
          const count = (failCounts.get(modelId) || 0) + 1;
          failCounts.set(modelId, count);
          const idx = Math.min(count - 1, BACKOFF.length - 1);
          failedUntil.set(modelId, Date.now() + BACKOFF[idx]);
          log(new Date().toISOString(), `MODEL STILL DOWN: ${modelId} (#${count}) — extended for ${BACKOFF[idx] / 1000}s`);
        }
      }
    }
  }, PROBE_INTERVAL);
  probeTimer.unref();
}

export function stopProbeLoop() {
  if (probeTimer) { clearInterval(probeTimer); probeTimer = null; }
}

File: internal/models/health.go
package models

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"sync"
	"time"
)

// Health timing mirrors src/models/health.mjs.
const (
	probeTimeout  = 8 * time.Second   // per-probe deadline
	exhaustLockMs = 120 * time.Second // 429 rate-limit lock
	degradeLockMs = 60 * time.Second  // degraded (slow/empty/sse-timeout) lock
)

var (
	// probeInterval is the recovery-probe cadence (60s). A variable so tests
	// can shorten it.
	probeInterval = 60 * time.Second
	// backoff is the consecutive-failure cooldown ladder:
	// 30s, 1m, 2m, 5m, 10m.
	backoff = []time.Duration{30 * time.Second, 60 * time.Second, 120 * time.Second, 300 * time.Second, 600 * time.Second}
)

// Health tracks per-model failure locks. Ported from src/models/health.mjs.
// failedUntil holds expiry timestamps in unix millis, failCounts tracks
// consecutive failures that drive the BACKOFF ladder.
type Health struct {
	mu          sync.Mutex
	failedUntil map[string]int64
	failCounts  map[string]int
}

// NewHealth returns an empty tracker, all models healthy.
func NewHealth() *Health {
	return &Health{
		failedUntil: make(map[string]int64),
		failCounts:  make(map[string]int),
	}
}

// IsHealthy reports whether the model is not locked or its lock expired.
func (h *Health) IsHealthy(modelID string) bool {
	h.mu.Lock()
	defer h.mu.Unlock()
	until, ok := h.failedUntil[modelID]
	if !ok {
		return true
	}
	return time.Now().UnixMilli() > until
}

// HealthyModels returns only the currently healthy models.
func (h *Health) HealthyModels(models []Model) []Model {
	var out []Model
	for _, m := range models {
		if h.IsHealthy(m.ID) {
			out = append(out, m)
		}
	}
	return out
}

// MarkFailed locks a model after a 5xx. 4xx never locks, mirrors Node
// statusCode >= 500. Each consecutive failure steps up the BACKOFF ladder,
// capped at 600s.
func (h *Health) MarkFailed(modelID string, statusCode int) {
	if modelID == "" || statusCode < 500 {
		return
	}
	h.mu.Lock()
	count := h.failCounts[modelID] + 1
	h.failCounts[modelID] = count
	cooldown := backoff[ladderIndex(count)]
	h.failedUntil[modelID] = time.Now().UnixMilli() + cooldown.Milliseconds()
	h.mu.Unlock()
	slog.Info(fmt.Sprintf("MODEL UNHEALTHY: %s (%d, #%d), locked for %ds", modelID, statusCode, count, int(cooldown.Seconds())))
}

// MarkExhausted locks a 429 rate-limited model for 120s. Failure count
// is not changed (src/health.mjs markModelExhausted).
func (h *Health) MarkExhausted(modelID string) {
	if modelID == "" {
		return
	}
	h.mu.Lock()
	h.failedUntil[modelID] = time.Now().UnixMilli() + exhaustLockMs.Milliseconds()
	h.mu.Unlock()
	slog.Info(fmt.Sprintf("MODEL EXHAUSTED (429): %s, locked for %ds", modelID, int(exhaustLockMs.Seconds())))
}

// MarkDegraded locks a slow, empty or SSE-timeout model for 60s.
func (h *Health) MarkDegraded(modelID, reason string) {
	if modelID == "" {
		return
	}
	if reason == "" {
		reason = "degraded"
	}
	h.mu.Lock()
	h.failedUntil[modelID] = time.Now().UnixMilli() + degradeLockMs.Milliseconds()
	h.mu.Unlock()
	slog.Info(fmt.Sprintf("MODEL DEGRADED: %s (%s), locked for %ds", modelID, reason, int(degradeLockMs.Seconds())))
}

// ProbeLoop runs recovery probes until ctx is canceled. Every
// probeInterval it probes expired locks with POST /v1/messages (8s timeout,
// spoof headers and current WAF cookie). 200 clears the lock, anything else
// extends it. Cancel ctx to stop, there is no separate Stop method.
func (h *Health) ProbeLoop(ctx context.Context, client *http.Client, host string, port int, getHeaders func() map[string]string, getCookie func() string) {
	ticker := time.NewTicker(probeInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			h.probeOnce(ctx, client, host, port, getHeaders, getCookie)
		}
	}
}

// probeOnce finds expired locks and probes each one.
func (h *Health) probeOnce(ctx context.Context, client *http.Client, host string, port int, getHeaders func() map[string]string, getCookie func() string) {
	h.mu.Lock()
	if len(h.failedUntil) == 0 {
		h.mu.Unlock()
		return
	}
	now := time.Now().UnixMilli()
	var due []string
	for id, until := range h.failedUntil {
		if now > until {
			due = append(due, id)
		}
	}
	h.mu.Unlock()

	for _, id := range due {
		var headers map[string]string
		if getHeaders != nil {
			headers = getHeaders()
		}
		cookie := ""
		if getCookie != nil {
			cookie = getCookie()
		}
		if probeModel(ctx, client, host, port, id, cookie, headers) {
			h.clearLock(id)
		} else {
			h.extendLock(id)
		}
	}
}

func (h *Health) clearLock(modelID string) {
	h.mu.Lock()
	delete(h.failedUntil, modelID)
	delete(h.failCounts, modelID)
	h.mu.Unlock()
	slog.Info("MODEL RECOVERED: " + modelID)
}

// extendLock extends a still-failing lock with the next BACKOFF step.
// This is the probe path, separate from MarkFailed which is for live failures.
func (h *Health) extendLock(modelID string) {
	h.mu.Lock()
	count := h.failCounts[modelID] + 1
	h.failCounts[modelID] = count
	cooldown := backoff[ladderIndex(count)]
	h.failedUntil[modelID] = time.Now().UnixMilli() + cooldown.Milliseconds()
	h.mu.Unlock()
	slog.Warn(fmt.Sprintf("MODEL STILL DOWN: %s (#%d), extended for %ds", modelID, count, int(cooldown.Seconds())))
}

// ladderIndex clamps a 1-based count to the BACKOFF ladder.
func ladderIndex(count int) int {
	if count-1 >= len(backoff) {
		return len(backoff) - 1
	}
	return count - 1
}

// probeModel sends one recovery probe with POST /v1/messages, minimal
// payload, spoof headers and WAF cookie. Returns true only on 200. Uses
// a fresh 8s client that shares the caller's Transport.
func probeModel(ctx context.Context, client *http.Client, host string, port int, modelID, cookie string, headers map[string]string) bool {
	body, err := json.Marshal(map[string]any{
		"model":      modelID,
		"max_tokens": 1,
		"stream":     false,
		"messages":   []map[string]string{{"role": "user", "content": "."}},
	})
	if err != nil {
		return false
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, upstreamURL(host, port, "/v1/messages"), bytes.NewReader(body))
	if err != nil {
		return false
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	req.Header.Set("Content-Type", "application/json")
	if cookie != "" {
		req.Header.Set("Cookie", cookie)
	}
	resp, err := clientWithTimeout(client, probeTimeout).Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	_, _ = io.Copy(io.Discard, resp.Body)
	return resp.StatusCode == http.StatusOK
}
